src/v/storage/batch_cache.cc

```diff
@@ -370,6 +370,22 @@ void batch_cache_index::truncate(model::offset offset) {
     }
 }
 
+void batch_cache_index::prefix_truncate(model::offset offset) {
+    lock_guard lk(*this);
+    if (auto it = _index.lower_bound(offset); it != _index.end()) {
+        if (it == _index.begin()) {
+            return;
+        }
+
+        std::for_each(
+          _index.begin(), std::prev(it), [this](index_type::value_type& e) {
+              _cache->evict(std::move(e.second.range()));
+          });
+
+        _index.erase(_index.begin(), it);
```

Comment:
it might be worth mentioning how the right side of the iterator intervals are chosen. it looks like a different range is erased than is iterated above, and it's not clear why.